core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -468,21 +484,50 @@ private void handleCacheStreamException(IOException e) {
   }
 
   /**
-   * Only updates {@link #mCurrentCacheStream}, {@link #mCurrentBlockInStream} and
-   * {@link #mStreamBlockId} to be in sync with the current block (i.e. {@link #getCurrentBlockId()}
-   * ). If this method is called multiple times, the subsequent invokes become no-op. Call this
-   * function every read and seek unless you are sure about the block streams are up-to-date.
+   * Checks whether the current block stream has remaining bytes, if not, advances to the next block
+   * stream. Otherwise, this is a no-op.
+   *
+   * This should only be used in streaming read APIs. For APIs like {@link #seek(long)},
+   * {@link #skip(long)}, the block stream should be updated according to the current read position,
+   * use {@link #updateStreamsBasedOnPosition()} instead.
    */
   private void updateStreams() throws IOException {
+    checkCacheStreamInSync();
+    if (shouldUpdateStreams()) {
+      mBlockIndex++;
```

Comment:
do we rely on `mBlockIndex++` to be `0` when `mBlockIndex = UNINITIALIZED_BLOCK_INDEX`? if yes, this is tricky, as I would imagine `UNINITIALIZED_BLOCK_INDEX` can be any negative value, which may break this assumption

Reply:
yes, but if we relax this assumption, there would be extra `if` statements or boolean checks which are not necessary.

Reply:
I've updated the PR to relax this assumption.